pkg/sql/schemachanger/doc.go

```diff
@@ -0,0 +1,169 @@
+// remain accessible to concurrent queries while they undergo schema changes.
+// Online schema changes are enabled by upholding the 2-version invariant, which
+// necessarily implies that schema changes may require multiple transactions to
+// complete. Upholding correctness requires maintaining some atomicity and
+// isolation properties which in turn means that schema changes which fail must
+// be rolled back somehow. Schema changes may legitimately fail, typically when
+// adding a constraint on a nonempty table.
+//
+// In CockroachDB, both in the legacy and in the declarative schema changer, for
+// schema changes involving more than one transaction. this sequence of
+// subsequent transactions is driven by a job; of type SCHEMA_CHANGE or
+// TYPEDESC_SCHEMA_CHANGE for the legacy schema changer, NEW_SCHEMA_CHANGE for
+// the declarative schema changer. In both cases the schema change gets rolled
+// back when the job's Resume method encounters an error, which eventually
+// transitions the job to the `failed` terminal state. Both schema changers also
+// have in common the fact that they store the current state and targeted
+// end-state of an ongoing schema change inside the affected descriptor, but
+// here the similarities end: for the legacy schema changer, this state is
+// encoded in the descriptor protobuf's `mutations` slice; for the declarative
+// schema changer, it's `declarative_schema_changer_state`.
+//
+// # The element model
+//
+// The declarative schema changer models its persisted state in terms of
+// elements, which can be thought of as "things" on which a schema change can be
+// effected, typically leaves of a DDL statement's AST: e.g. a table's name is
+// modeled as an element called TableName, a column's default value expression
+// is modelled by ColumnDefaultExpression, and so forth.
+//
+// A schema change is define by setting target end-states to a set of elements.
+// These states are PUBLIC, ABSENT or TRANSIENT_ABSENT. This is best described
+// by example; `ALTER TABLE foo RENAME COLUMN x TO y`, assuming `foo` has
+// descriptor ID 123 and `x` has column ID 4, gets translated into:
+//  - ColumnName{DescID: 123, ColumnID: 4, Name: x} targets ABSENT,
+//  - ColumnName{DescID: 123, ColumnID: 4, Name: y} targets PUBLIC.
+//
+// When the schema change starts, those two elements are respectively in the
+// PUBLIC and ABSENT states. Performing the schema change means transitioning
+// these to their target states. This may include intermediate states, this
+// isn't the case for ColumnName but a Column transitions through DELETE_ONLY,
+// WRITE_ONLY, etc. because of the need to uphold the 2-version invariant.
+// Elements and states are defined in `scpb`. The mapping of the descriptor
+// model (i.e. what's in system.descriptor) to the element model (which is what
+// the declarative schema changer understands) is defined and implemented in
```

Comment:
Do you think it's helpful for the reader if we point out the fact that "one descriptor *decomposes* to a set of elements, and those elements *belong to* this one descriptor, effectively forming a one-to-many relationship?"

Reply:
Thanks for the suggestion but I think it's a bit of a distraction to mention that. IIRC the scdecomp and scpb packages do document that. If they don't, it belongs there.